File: app/modules/dashboard_auth/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.core.auth.dashboard_access import DashboardPermission, DashboardRole
from app.core.auth.dashboard_mode import DashboardAuthMode
from app.core.auth.step_up import StepUpMethod
from app.modules.shared.schemas import DashboardModel


class DashboardUserRoleSummary(DashboardModel):
    id: str
    slug: str
    name: str
    kind: str


class DashboardSessionUser(DashboardModel):
    id: str
    username: str
    display_name: str | None = None
    role: DashboardUserRoleSummary


LoginProviderKind = Literal["password", "trusted_header", "oidc"]
#: ``dashboard_settings.local_login_policy`` on the wire (PLAN §4.6).
LocalLoginPolicyValue = Literal["enabled", "admins_only", "break_glass_only"]


class DashboardLoginProvider(DashboardModel):
    kind: LoginProviderKind
    provider_key: str = "default"
    label: str
    login_url: str | None = None


class DashboardPendingArrival(DashboardModel):
    """What a browser the identity resolver refused may be told about its own arrival.

    ``provider`` is the row's operator-chosen label -- the same public string
    the sign-in button carries -- and ``reference`` the masked form of the
    address the identity provider asserted, which is a lossy projection of
    something this browser itself presented. Neither says whether any account
    exists, and no subject, group, claim or address in clear appears here.
    """

    provider: str
    reference: str


class DashboardLoginHint(DashboardModel):
    """Login-screen hints served to unauthenticated clients too; never carries a username.

    ``pending_identity`` is true when the request carried a provider-asserted
    identity that has no account here (refused or not provisioned): the client
    shows "your account is not ready yet" instead of a login form.
    """

    username_field: Literal["hidden", "shown"]
    providers: list[DashboardLoginProvider]
    #: Whether the local password form is shown, collapsed behind a link
    #: (``admins_only``) or reachable only at ``/login?local=1``
    #: (``break_glass_only``). The account name is never sent here.
    local_login: LocalLoginPolicyValue = "enabled"
    pending_identity: bool = False
    #: Set only from the sealed marker an OIDC refusal left for this browser,
    #: and from nothing else -- not from the active providers, not from any
    #: account lookup -- so no caller can ask for the block of an address of
    #: their choosing. A reverse-proxy refusal keeps the bare boolean above.
    pending_arrival: DashboardPendingArrival | None = None


class DashboardAccessSummary(DashboardModel):
    """Team-size facts for `users:manage` holders only (drives the frontend disclosure tier)."""

    users_total: int
    users_active: int
    users_invited: int
    users_disabled: int
    pending_invites: int
    non_admin_users: int
    custom_roles: int
    providers_enabled: list[str]
    role_mappings: int
    scim_tokens: int
    audit_sinks: int
    local_login_policy: LocalLoginPolicyValue


class DashboardStepUpState(DashboardModel):
    """Whether the account has recently re-verified for sensitive changes, and how it can.

    ``methods`` lists every factor the account must present to ``/step-up``
    (``password`` and/or ``totp``); empty means it cannot step up until it
    enrols two-factor or sets a local password.
    """

    verified_at: int | None = None
    expires_at: int | None = None
    methods: list[StepUpMethod] = Field(default_factory=list)


class DashboardAuthSessionResponse(DashboardModel):
    authenticated: bool
    password_required: bool
    #: An active account holds a password (drives the Password card; accounts that
    #: sign in through a provider alone do not count).
    local_password_configured: bool = False
    totp_required_on_login: bool
    totp_configured: bool
    bootstrap_required: bool = False
    bootstrap_token_configured: bool = False
    auth_mode: DashboardAuthMode = DashboardAuthMode.STANDARD
    password_management_enabled: bool = True
    password_session_active: bool = False
    #: Coarse wire role; every signed-in account is ``admin`` here, ``user.role`` carries the real one.
    role: DashboardRole = DashboardRole.ADMIN
    #: Legacy ``read``/``write`` aliases followed by ``<permission>:<scope>`` entries.
    permissions: list[str] = Field(
        default_factory=lambda: [DashboardPermission.READ.value, DashboardPermission.WRITE.value]
    )
    guest_access_enabled: bool = False
    guest_password_required: bool = False
    user: DashboardSessionUser | None = None
    auth_method: str | None = None
    must_change_password: bool = False
    totp_enrollment_required: bool = False
    login: DashboardLoginHint | None = None
    access_summary: DashboardAccessSummary | None = None
    assignable_role_ids: list[str] = Field(default_factory=list)
    #: Present for signed-in accounts only; principals without an account have nothing to re-verify.
    step_up: DashboardStepUpState | None = None
    #: True when the session was minted for an account carrying the
    #: break-glass designation, so the header can show the emergency pill.
    break_glass_session: bool = False


class DashboardMeResponse(DashboardModel):
    id: str
    username: str
    display_name: str | None = None
    email: str | None = None
    role: DashboardUserRoleSummary
    auth_method: str | None = None
    totp_configured: bool
    must_change_password: bool


class TotpSetupStartResponse(DashboardModel):
    secret: str
    otpauth_uri: str
    qr_svg_data_uri: str


class TotpSetupConfirmRequest(DashboardModel):
    secret: str
    code: str


class TotpVerifyRequest(DashboardModel):
    code: str


class StepUpRequest(DashboardModel):
    """The factors offered to ``POST /api/dashboard-auth/step-up``; which are needed depends on the account."""

    password: str | None = None
    code: str | None = Field(default=None, max_length=16)


class StepUpResponse(DashboardModel):
    verified_at: int
    expires_at: int


class OidcStartResponse(DashboardModel):
    """Where the dashboard must send the browser to begin a signed-in OIDC round trip.

    The URL is returned rather than served as a redirect because the caller is
    the dashboard's own script, which opens it in a popup (falling back to the
    current tab) and waits for the callback to land back on the settings page.
    """

    authorization_url: str


class PasswordSetupRequest(DashboardModel):
    password: str
    bootstrap_token: str | None = None


class PasswordLoginRequest(DashboardModel):
    #: Optional on single-user installs; required once more than one account holds a password.
    username: str | None = Field(default=None, max_length=64)
    password: str


class GuestLoginRequest(DashboardModel):
    password: str | None = None


class GuestPasswordSetRequest(DashboardModel):
    password: str


class PasswordChangeRequest(DashboardModel):
    current_password: str
    new_password: str


class PasswordRemoveRequest(DashboardModel):
    password: str


class InviteDescriptionResponse(DashboardModel):
    """What the acceptance screen may learn from a valid invite token."""

    role_name: str
    inviter_display_name: str | None = None
    suggested_username: str
    username_locked: bool
    expires_at: datetime


class InviteAcceptRequest(DashboardModel):
    token: str = Field(max_length=256)
    username: str | None = Field(default=None, max_length=64)
    password: str
    display_name: str | None = Field(default=None, max_length=128)

File: frontend/src/features/auth/schemas.ts
import { z } from "zod";

// Mirrors backend `_MAX_PASSWORD_BYTES` in `app/modules/dashboard_auth/api.py`.
// bcrypt only hashes the first 72 bytes of input; the backend rejects
// anything longer with HTTP 422 `password_too_long`. Validate the same
// budget on the client so users see the failure inline instead of after a
// round-trip. Multi-byte characters (e.g. emoji) consume their UTF-8 byte
// count, not the visible character count.
export const MAX_DASHBOARD_PASSWORD_BYTES = 72;

const dashboardPasswordByteLength = (value: string): number => new TextEncoder().encode(value).length;

const PASSWORD_REQUIRED_MESSAGE = "settings.password.validation.required";

const dashboardPasswordSchema = z
  .string()
  .min(8, {
    message: "settings.password.validation.minLength",
  })
  .refine((value) => dashboardPasswordByteLength(value) <= MAX_DASHBOARD_PASSWORD_BYTES, {
    message: "settings.password.validation.maxByteLength",
  });

export const DashboardAuthModeSchema = z.enum(["standard", "trusted_header", "disabled"]);
export const DashboardRoleSchema = z.enum(["admin", "guest"]);
export const DashboardPermissionSchema = z.enum(["read", "write"]);

// Fine-grained permissions (`<resource>:<action>`); mirrors `Permission` in
// `app/core/auth/dashboard_access.py`. The session lists each grant as
// `<permission>:<scope>` after the coarse aliases.
export const PermissionSchema = z.enum([
  "dashboard:read",
  "accounts:read",
  "accounts:write",
  "accounts:export",
  "api_keys:read",
  "api_keys:write",
  "api_keys:assign",
  "ops:write",
  "security:write",
  "users:manage",
  "roles:manage",
  "conversations:read",
  "audit:read",
]);
export const PermissionScopeSchema = z.enum(["all", "own"]);

export const AuthSessionUserSchema = z.object({
  id: z.string(),
  username: z.string(),
  displayName: z.string().nullable().default(null),
  role: z.object({
    id: z.string(),
    slug: z.string(),
    name: z.string(),
    kind: z.string(),
  }),
});

export const LoginProviderSchema = z.object({
  kind: z.string(),
  providerKey: z.string().default("default"),
  label: z.string(),
  loginUrl: z.string().nullable().default(null),
});

// Who may still sign in with a local password (`dashboard_settings.local_login_policy`).
// The strict enum is what a *request* must satisfy: `updateSettings` takes
// `unknown`, so this schema is the only thing standing between a bad value and
// the wire, and a fallback there would silently send the most open policy of
// the three.
export const StrictLocalLoginPolicySchema = z.enum(["enabled", "admins_only", "break_glass_only"]);
// Responses get the fallback instead: a value this build does not know must not
// fail the whole session parse. Nothing may echo the fallen-back value into an
// update — see `buildSettingsUpdateRequest`.
export const LocalLoginPolicySchema = StrictLocalLoginPolicySchema.catch("enabled");

// What a browser the identity resolver refused may be told about its own
// arrival: the provider's public label (the same string its sign-in button
// carries) and the reference the server computed from the address the identity
// provider asserted. Both are the server's words; the client renders them and
// derives nothing. Absent for a reverse-proxy refusal and for an expired marker.
export const PendingArrivalSchema = z.object({
  provider: z.string(),
  reference: z.string(),
});

export const LoginHintSchema = z.object({
  usernameField: z.enum(["hidden", "shown"]).default("hidden"),
  providers: z
    .array(LoginProviderSchema)
    .default([{ kind: "password", providerKey: "default", label: "Password", loginUrl: null }]),
  localLogin: LocalLoginPolicySchema.default("enabled"),
  // The request carried a provider identity that has no account here yet.
  pendingIdentity: z.boolean().default(false),
  pendingArrival: PendingArrivalSchema.nullable().default(null),
});

// Team-size facts served only to `users:manage` holders; `null` for everyone
// else. Consumed exclusively by `resolveDisclosureTier`.
export const AccessSummarySchema = z.object({
  usersTotal: z.number().int().default(0),
  usersActive: z.number().int().default(0),
  usersInvited: z.number().int().default(0),
  usersDisabled: z.number().int().default(0),
  pendingInvites: z.number().int().default(0),
  nonAdminUsers: z.number().int().default(0),
  customRoles: z.number().int().default(0),
  providersEnabled: z.array(z.string()).default([]),
  roleMappings: z.number().int().default(0),
  scimTokens: z.number().int().default(0),
  auditSinks: z.number().int().default(0),
  localLoginPolicy: LocalLoginPolicySchema.default("enabled"),
});

// Step-up (re-verification for sensitive changes): when the account last
// re-verified, and which factors `/step-up` will ask for. Empty `methods`
// means the account must enrol two-factor or set a password first.
export const StepUpMethodSchema = z.enum(["password", "totp", "oidc"]);
export const StepUpStateSchema = z.object({
  verifiedAt: z.number().int().nullable().default(null),
  expiresAt: z.number().int().nullable().default(null),
  methods: z.array(StepUpMethodSchema).default([]),
});

// Where to send the browser to begin a signed-in round trip at the identity
// provider. The server builds the URL from the stored configuration; the app
// only follows it.
export const OidcStartResponseSchema = z.object({
  authorizationUrl: z.string(),
});

export const AuthSessionSchema = z.object({
  authenticated: z.boolean(),
  passwordRequired: z.boolean(),
  // An active account holds a password (proxy-created accounts do not count).
  localPasswordConfigured: z.boolean().default(false),
  totpRequiredOnLogin: z.boolean(),
  totpConfigured: z.boolean(),
  bootstrapRequired: z.boolean().optional().default(false),
  bootstrapTokenConfigured: z.boolean().optional().default(false),
  authMode: DashboardAuthModeSchema.default("standard"),
  passwordManagementEnabled: z.boolean().default(true),
  passwordSessionActive: z.boolean().default(false),
  // Least-privilege defaults: a response that omits these fields must never
  // be treated as an admin. `permissions` accepts any string so that future
  // fine-grained values (e.g. `accounts:export`) do not reject the session.
  role: DashboardRoleSchema.default("guest"),
  permissions: z.array(z.string()).default([]),
  guestAccessEnabled: z.boolean().default(false),
  guestPasswordRequired: z.boolean().default(false),
  // Account fields (user-login-and-session-v2). Every default is the
  // "no account / individual install" shape so older payloads parse unchanged.
  user: AuthSessionUserSchema.nullable().default(null),
  authMethod: z.string().nullable().default(null),
  mustChangePassword: z.boolean().default(false),
  totpEnrollmentRequired: z.boolean().default(false),
  login: LoginHintSchema.nullable()
    .default(null)
    .transform((value) => value ?? LoginHintSchema.parse({})),
  accessSummary: AccessSummarySchema.nullable().default(null),
  assignableRoleIds: z.array(z.string()).default([]),
  stepUp: StepUpStateSchema.nullable().default(null),
  // The session was minted for an account carrying the break-glass
  // designation; the header says so, nothing else changes.
  breakGlassSession: z.boolean().default(false),
});

// Mirrors the backend username rule (case-folded on the server).
const USERNAME_PATTERN = /^[a-z0-9._-]{1,64}$/i;
const usernameSchema = z
  .string()
  .trim()
  .min(1, { message: "auth.invite.validation.usernameRequired" })
  .max(64, { message: "auth.invite.validation.usernameTooLong" })
  .regex(USERNAME_PATTERN, { message: "auth.invite.validation.usernameFormat" });

export const LoginRequestSchema = z.object({
  username: z.string().trim().max(64, { message: "auth.invite.validation.usernameTooLong" }).optional(),
  password: z.string().min(1),
});

export const InviteDescriptionSchema = z.object({
  roleName: z.string(),
  inviterDisplayName: z.string().nullable().default(null),
  suggestedUsername: z.string(),
  usernameLocked: z.boolean(),
  expiresAt: z.string(),
});

export const InviteAcceptRequestSchema = z.object({
  token: z.string().min(1),
  username: usernameSchema,
  password: dashboardPasswordSchema,
  displayName: z.string().trim().max(128).optional(),
});

export const GuestLoginRequestSchema = z.object({
  password: z.string().optional(),
});

export const PasswordSetupRequestSchema = z.object({
  password: dashboardPasswordSchema,
  bootstrapToken: z.string().optional(),
});

export const GuestPasswordSetRequestSchema = z.object({
  password: dashboardPasswordSchema,
});

export const PasswordChangeRequestSchema = z.object({
  currentPassword: z.string().min(1, { message: PASSWORD_REQUIRED_MESSAGE }),
  newPassword: dashboardPasswordSchema,
});

export const PasswordRemoveRequestSchema = z.object({
  password: z.string().min(1, { message: PASSWORD_REQUIRED_MESSAGE }),
});

export const TotpVerifyRequestSchema = z.object({
  code: z.string().length(6, "settings.totp.validation.codeLength"),
});

export const TotpSetupConfirmRequestSchema = z.object({
  secret: z.string().min(1),
  code: z.string().min(6).max(6),
});

export const TotpSetupStartResponseSchema = z.object({
  secret: z.string(),
  otpauthUri: z.string(),
  qrSvgDataUri: z.string(),
});

export const StatusResponseSchema = z.object({
  status: z.string(),
});

export const StepUpRequestSchema = z.object({
  password: z.string().min(1).optional(),
  code: z.string().length(6, "settings.totp.validation.codeLength").optional(),
});

export const StepUpResponseSchema = z.object({
  verifiedAt: z.number().int(),
  expiresAt: z.number().int(),
});

export type AuthSession = z.infer<typeof AuthSessionSchema>;
export type AuthSessionUser = z.infer<typeof AuthSessionUserSchema>;
export type LoginHint = z.infer<typeof LoginHintSchema>;
export type LoginProvider = z.infer<typeof LoginProviderSchema>;
export type PendingArrival = z.infer<typeof PendingArrivalSchema>;
export type LocalLoginPolicy = z.infer<typeof LocalLoginPolicySchema>;
export type AccessSummary = z.infer<typeof AccessSummarySchema>;
export type Permission = z.infer<typeof PermissionSchema>;
export type PermissionScope = z.infer<typeof PermissionScopeSchema>;
export type InviteDescription = z.infer<typeof InviteDescriptionSchema>;
export type InviteAcceptRequest = z.infer<typeof InviteAcceptRequestSchema>;
export type DashboardAuthMode = z.infer<typeof DashboardAuthModeSchema>;
export type DashboardRole = z.infer<typeof DashboardRoleSchema>;
export type DashboardPermission = z.infer<typeof DashboardPermissionSchema>;
export type LoginRequest = z.infer<typeof LoginRequestSchema>;
export type GuestLoginRequest = z.infer<typeof GuestLoginRequestSchema>;
export type PasswordSetupRequest = z.infer<typeof PasswordSetupRequestSchema>;
export type GuestPasswordSetRequest = z.infer<typeof GuestPasswordSetRequestSchema>;
export type PasswordChangeRequest = z.infer<typeof PasswordChangeRequestSchema>;
export type PasswordRemoveRequest = z.infer<typeof PasswordRemoveRequestSchema>;
export type TotpVerifyRequest = z.infer<typeof TotpVerifyRequestSchema>;
export type TotpSetupConfirmRequest = z.infer<typeof TotpSetupConfirmRequestSchema>;
export type TotpSetupStartResponse = z.infer<typeof TotpSetupStartResponseSchema>;
export type StatusResponse = z.infer<typeof StatusResponseSchema>;
export type StepUpState = z.infer<typeof StepUpStateSchema>;
export type StepUpRequest = z.infer<typeof StepUpRequestSchema>;
export type OidcStartResponse = z.infer<typeof OidcStartResponseSchema>;

export function getFirstZodIssueMessage(error: unknown): string | null {
  if (!(error instanceof z.ZodError)) {
    return null;
  }

  const [firstIssue] = error.issues;
  return typeof firstIssue?.message === "string" ? firstIssue.message : null;
}
